### DISCUSSION TITLE
Bunch of questions if my day to day workflow can be supported

### DISCUSSION THREAD

**Discussion by zoiman** [ttscoff/bunch-docs#280] [General]
I mainly want to use bunch to switch to a project context out of pure chaos which should be 
restored after I quit the bunch.

Before I switch to my project context I have multiple chrome and firefox instances open and iTerm
instances and a bunch of other apps.

This is what should happen when I start the project bunch:

- hide all open browser instances ( I think this is possible)
- open Webstorm with a specific project (not checked but I think this is possible)
- open 2 new chrome windows with a specific chrome profile (no idea if this is possible)
- open firefox with some tabs in some specific containers (https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/)  - dont think this is possible
- open iTerm and cd to specific location ( think this should be possible)
- place the windows on 2 monitors accordingly, this should be possible with Moom however I am using RectanglePro maybe this is also possible using keystrokes?

If this is possible, it already great however I often have to switch between contexts back and forth without loosing the “state” (browser tabs, window positioning).
So for example:
I have multiple instances of chrome/ firefox, iTerm open doing “normal” work stuff then I want to work on a project and start the project bunch which should only display the needed instances of apps as described above.( and start them)
Then I want to quickly switch to the state before to do some “normal” stuff.
When I now switch back to the project everything should look the same as I left it (same tabs open etc)

But when I am done working on the project for the day I want to really quit everything project related. So there needs to be a difference between switching and quitting.

Is this somehow possible?

I was thinking to use different Desktops where everything of one project bunch is launched on a different Desktop and the quick switching is done by changing the Desktop but I did not find anything in the docs where I can launch an app in a different Desktop

**Comment by ttscoff** [comment]
On 10 Dec 2022, at 9:47, Simon Plutz wrote:

> - open Webstorm with a specific project (not checked but I think this 
> is possible)

Opening files/projects depends on how well a given application responds 
to open commands from macOS. I don't have Webstorm installed to test 
this, but should be as easy as

```
Webstorm
- /path/to/project
```

> - open 2 new chrome windows with a specific chrome profile (no idea if 
> this is possible)

	chrome[Profile Name]://url

> - open firefox with some tabs in some specific containers 
> (https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/) 
>  - dont think this is possible

No, Firefox offers no support for externally controlling containers, 
especially those provided by an extension.

> - open iTerm and cd to specific location ( think this should be 
> possible)

iTerm responds to open commands, AppleScript commands, and you can also 
send keystrokes directly to it, e.g.

```
iTerm
- {"cd ~/my_dir" return}
```

> - place the windows on 2 monitors accordingly, this should be possible 
> with Moom however I am using RectanglePro maybe this is also possible 
> using keystrokes?

You can send global keystrokes by just using curly brackets on a line by 
themselves. Results can be mixed. Does RectanglePro have any AppleScript 
library?

>

> So for example:
> I have multiple instances of chrome/ firefox, iTerm open doing 
> “normal” work stuff then I want to work on a project and start the 
> project bunch which should only display the needed instances of apps 
> as described above.( and start them)
> Then I want to quickly switch to the state before to do some 
> “normal” stuff.
> When I now switch back to the project everything should look the same 
> as I left it (same tabs open etc)
>
> But when I am done working on the project for the day I want to really 
> quit everything project related. So there needs to be a difference 
> between switching and quitting.

Not sure I understand the question fully, but you could easily put an 
[optional 
snippet](https://bunchapp.co/docs/bunch-files/interactivity/optional-snippets/) 
in a Bunch that, if confirmed, quit the applications, and if canceled 
nothing would happen. Just add a snippet to the Bunch and call it like 
this:

```
!<<#Quit ?"Quit other apps" // Ask when closing

// Open apps
// Use %AppName to not automatically quit when closing
%Webstorm
%Messages
%Whatever
___
#[Quit]
!Webstorm
!Messages
!Whatever
```

Does that make sense? You could also have separate snippets that were 
triggered based on a variable like time of day or other trigger. 
https://bunchapp.co/docs/bunch-files/logic/

>
> Is this somehow possible?
>
> I was thinking to use different Desktops where everything of one 
> project bunch is launched on a different Desktop and the quick 
> switching is done by changing the Desktop but I did not find anything 
> in the docs where I can launch an app in a different Desktop

macOS offers no interface for launching apps in Spaces (or in the new 
Stage Manager, and no window manager I know of can do it, either.

**Comment by zoiman** [comment]
thank you for the detailed explanation, this already helps me alot.
I am still not sure about switching contexts without loosing the state. My main problem with my current workflow is, that I just have too many apps and instances of apps open where only a small amount of them concern me for a specific task, but I can't close the others since I would loose some kind of state.

Maybe I try the simplest example for a better understanding.

I have 2 chrome instances open where each one has multiple tabs some of them have unsent forms for example. 
Now I want to work on a project so I start the project bunch.
This should completely hide those chrome instances and start a new one with some tabs. 
Then I also add some more tabs, start filling out some forms etc. 
Then I want to go back to the previous state so that the current chrome instances are hidden and the other pop up again.


So in summary, all my project bunches should when started hide everthing and when left hide the created apps in that bunch and unhide the ones which were hidden before.
And with the optional you described above I would also add the option to quit which really quits these apps of a bunch.

Does that make sense?

**Reply by ttscoff** [reply]
Bunch cannot close tabs or windows, it can only quit apps. So yes, you could completely quit Chrome, but you can't close just the windows/tabs opened by a Bunch. It also cannot unhide apps that were hidden with a command like @@ or @App focusing. That's likely beyond the scope of Bunch.

**Comment by zoiman** [comment]
So what happens when I am on a new space an start a bunch? Are those apps started in the space I am currently in?

**Reply by ttscoff** [reply]
Yes.

**Comment by dcrossleyau** [comment]
This enables opening stuff on each relevant Desktop Space:

```
---
title: D:4 Doc Bunch
---

/*
NOTE: Configure keyboard shortcuts to switch desktops in Mission Control.
Preferences > Keyboard > Shortcuts > Mission Control

NOTE: Need an application window somewhere on the respective Spaces.
The new window dimensions will be the same as the most-recently used window.

Docs:
https://bunchapp.co/docs/bunch-files/keystrokes/#sending-keystrokes-and-combos
*/

// Move to Desktop Space
{^4}

%Firefox^
- {@n "https://github.com/ttscoff/bunch" return}
- {@t "https://bunchapp.co/docs" return}
```
